scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/constraints/TimeRangeExpression.java

```diff
@@ -30,91 +30,78 @@ public class TimeRangeExpression {
-      if(minTimepoint.isPresent()&& maxTimepoint.isPresent()) {
+      var minTimepoint = inter.minTimePoint();
+      var maxTimepoint = inter.maxTimePoint();
+      if(minTimepoint.isPresent() && maxTimepoint.isPresent()) {
         final var anchorActSearch = new ActivityExpression.Builder()
             .basedOn(actTemplate)
             .startsIn(Window.between(
                 Duration.max(Duration.ZERO, minTimepoint.get()),
                 Window.Inclusivity.Inclusive,
                 maxTimepoint.get(),
                 Window.Inclusivity.Exclusive)).build();
-        final var anchorActs = plan.find(anchorActSearch);
+        final var anchorActs = plan.find(anchorActSearch, simulationResults);
         for (var anchorAct : anchorActs) {
           var endInclusivity = Window.Inclusivity.Exclusive;
           if(anchorAct.getDuration().isZero()){
             endInclusivity = Window.Inclusivity.Inclusive;
           }
           actTw.add(Window.between(anchorAct.getStartTime(), Window.Inclusivity.Inclusive, anchorAct.getEndTime(), endInclusivity));
         }
-        inter = actTw;
       }
+      inter.intersectWith(actTw);
     }
 
     for (var otherExpr : timeRangeExpressions) {
-      Windows windowsState = otherExpr.computeRange(plan, domain);
+      if(inter.isEmpty()) break;
+      Windows windowsState = otherExpr.computeRange(simulationResults, plan, domain);
       inter.intersectWith(windowsState);
-      if (inter.isEmpty()) {
-        break;
-      }
     }
 
     for (var expr : stateExpr) {
-      Windows windowsState = expr.findWindows(plan, domain);
+      if(inter.isEmpty()) break;
+      final var domainOfInter = Window.between(inter.minTimePoint().get(), inter.maxTimePoint().get());
+      Windows windowsState = expr.evaluate(simulationResults, domainOfInter, null);
       inter.intersectWith(windowsState);
-      if (inter.isEmpty()) {
-        break;
-      }
     }
 
     for (var constState : constantsStates) {
-      Windows windowsState = new Windows(toOO(constState.getTimeline(domain).keySet().stream().toList()));
-      inter.intersectWith(windowsState);
-      if (inter.isEmpty()) {
-        break;
-      }
+      if(inter.isEmpty()) break;
+      final var domainOfInter = Window.between(inter.minTimePoint().get(), inter.maxTimePoint().get());
+      final var changePoints = simulationResults.discreteProfiles.get(constState).changePoints(domainOfInter);
+      final var timeline = new Windows();
+      final var container = new ArrayList<Window>();
+      changePoints.iterator().forEachRemaining(container::add);
+      container.stream().reduce((a, b) -> {
+        timeline.add(Window.window(a.start, Window.Inclusivity.Exclusive, b.start, Window.Inclusivity.Exclusive));
+        return b;
+      });
+      inter.intersectWith(timeline);
     }
+
     for (var filterOrtransform : filtersAndTransformers) {
+      if(inter.isEmpty()) break;
       if (filterOrtransform instanceof TimeWindowsFilter) {
```

Comment:
Here's another place where we could use "pattern matching instanceof" instead of casting:
```suggestion
      if (filterOrtransform instanceof TimeWindowsFilter f) {
```